### PR TITLE
feat: ng add of packages will run depending of project type

### PR DESCRIPTION
**Pull request by mrednic-1A**
## Related issues

- :bug: Fixes #(issue)
- :rocket: Feature #(issue)

### CODE REVIEW

File: packages/@o3r/components/schematics/ng-add/index.ts [hunk 1]
```diff
@@ -33,10 +34,11 @@ export function ngAdd(options: NgAddSchematicsSchema): Rule {
         depsInfo.o3rPeerDeps = [...depsInfo.o3rPeerDeps , '@o3r/extractors'];
       }
       const dependencyType = getProjectDepType(tree);
+      const workingDirectory = getProjectRootDir(tree, options.projectName);
       const rule = chain([
         removePackages(['@otter/components']),
-        ngAddPackages(depsInfo.o3rPeerDeps, { skipConfirmation: true, version: depsInfo.packageVersion, parentPackageInfo: depsInfo.packageName, dependencyType }),
-        ngAddPeerDependencyPackages(['chokidar'], packageJsonPath, NodeDependencyType.Dev, options, '@o3r/components - install builder dependency'),
+        ngAddPackages(depsInfo.o3rPeerDeps, { skipConfirmation: true, version: depsInfo.packageVersion, parentPackageInfo: depsInfo.packageName, dependencyType, workingDirectory }),
+        ngAddPeerDependencyPackages(['chokidar'], packageJsonPath, NodeDependencyType.Dev, {...options, workingDirectory, skipNgAddSchemticRun: true}, '@o3r/components - install builder dependency'),
```
Comment (matthieu-crouzet) [inline packages/@o3r/components/schematics/ng-add/index.ts:41]:
```suggestion
        ngAddPeerDependencyPackages(['chokidar'], packageJsonPath, NodeDependencyType.Dev, {...options, workingDirectory, skipNgAddSchematicRun: true}, '@o3r/components - install builder dependency'),
```

File: packages/@o3r/core/schematics/rule-factories/additional-modules/index.ts [hunk 2]
```diff
@@ -34,7 +34,7 @@ export function updateAdditionalModules(options: { projectName?: string | null |
     const generatorDependencies = [ngrxStoreDevtoolsDep];
 
     try {
-      return ngAddPeerDependencyPackages(generatorDependencies, packageJsonPath, type, options);
+      return ngAddPeerDependencyPackages(generatorDependencies, packageJsonPath, type, {...options, skipNgAddSchemticRun: true});
```
Comment (matthieu-crouzet) [inline packages/@o3r/core/schematics/rule-factories/additional-modules/index.ts:37]:
```suggestion
      return ngAddPeerDependencyPackages(generatorDependencies, packageJsonPath, type, {...options, skipNgAddSchematicRun: true});
```

File: packages/@o3r/core/schematics/rule-factories/store/index.ts [hunk 3]
```diff
@@ -53,7 +53,7 @@ export function updateStore(options: { projectName?: string | undefined}, projec
 
     return () => {
       try {
-        return ngAddPeerDependencyPackages(dependenciesList, packageJsonPath, type, options)(tree, context);
+        return ngAddPeerDependencyPackages(dependenciesList, packageJsonPath, type, {...options, skipNgAddSchemticRun: true})(tree, context);
```
Comment (matthieu-crouzet) [inline packages/@o3r/core/schematics/rule-factories/store/index.ts:56]:
```suggestion
        return ngAddPeerDependencyPackages(dependenciesList, packageJsonPath, type, {...options, skipNgAddSchematicRun: true})(tree, context);
```

File: packages/@o3r/eslint-config-otter/schematics/ng-add/index.ts [hunk 4]
```diff
@@ -12,20 +12,21 @@ export function ngAdd(options: NgAddSchematicsSchema): Rule {
...
       return () => chain([
         removePackages(['@otter/eslint-config-otter', '@otter/eslint-plugin']),
-        ngAddPackages(depsInfo.o3rPeerDeps, {skipConfirmation: true, version: depsInfo.packageVersion, parentPackageInfo: depsInfo.packageName, dependencyType}),
+        ngAddPackages(depsInfo.o3rPeerDeps, {skipConfirmation: true, version: depsInfo.packageVersion, parentPackageInfo: depsInfo.packageName, dependencyType, workingDirectory}),
         ngAddPeerDependencyPackages(
           ['eslint', '@angular-eslint/builder', '@typescript-eslint/parser', '@typescript-eslint/eslint-plugin', 'eslint-plugin-jsdoc', 'eslint-plugin-prefer-arrow', 'eslint-plugin-unicorn'],
           path.resolve(__dirname, '..', '..', 'package.json'),
           NodeDependencyType.Dev,
-          options,
+          {...options, workingDirectory, skipNgAddSchemticRun: true},
```
Comment (matthieu-crouzet) [inline packages/@o3r/eslint-config-otter/schematics/ng-add/index.ts:29]:
```suggestion
          {...options, workingDirectory, skipNgAddSchematicRun: true},
```

File: packages/@o3r/localization/schematics/localization-base/index.ts [hunk 5]
```diff
@@ -446,8 +447,9 @@ export function updateLocalization(options: { projectName?: string | null | unde
   const addDependencies: Rule = (tree: Tree, context: SchematicContext) => {
     const type: NodeDependencyType = getProjectDepType(tree);
     const generatorDependencies = [ngxTranslateCoreDep, intlMessageFormatDep, formatjsIntlNumberformatDep, angularCdkDep];
+    const workingDirectory = getProjectRootDir(tree, options.projectName);
     try {
-      return ngAddPeerDependencyPackages(generatorDependencies, packageJsonPath, type, options)(tree, context);
+      return ngAddPeerDependencyPackages(generatorDependencies, packageJsonPath, type, {...options, workingDirectory, skipNgAddSchemticRun: true})(tree, context);
```
Comment (matthieu-crouzet) [inline packages/@o3r/localization/schematics/localization-base/index.ts:452]:
```suggestion
      return ngAddPeerDependencyPackages(generatorDependencies, packageJsonPath, type, {...options, workingDirectory, skipNgAddSchematicRun: true})(tree, context);
```

File: packages/@o3r/rules-engine/schematics/ng-add/index.ts [hunk 6]
```diff
@@ -43,8 +45,8 @@ export function ngAdd(options: NgAddSchematicsSchema): Rule {
           }
         }),
         removePackages(['@otter/rules-engine', '@otter/rules-engine-core']),
-        ngAddPeerDependencyPackages(['jsonpath-plus'], packageJsonPath, dependencyType, options, '@o3r/rules-engine - install builder dependency'),
-        ngAddPackages(depsInfo.o3rPeerDeps, { skipConfirmation: true, version: depsInfo.packageVersion, parentPackageInfo: depsInfo.packageName, dependencyType }),
+        ngAddPeerDependencyPackages(['jsonpath-plus'], packageJsonPath, dependencyType, {...options, workingDirectory, skipNgAddSchemticRun: true}, '@o3r/rules-engine - install builder dependency'),
```
Comment (matthieu-crouzet) [inline packages/@o3r/rules-engine/schematics/ng-add/index.ts:48]:
```suggestion
        ngAddPeerDependencyPackages(['jsonpath-plus'], packageJsonPath, dependencyType, {...options, workingDirectory, skipNgAddSchematicRun: true}, '@o3r/rules-engine - install builder dependency'),
```

File: packages/@o3r/schematics/src/rule-factories/ng-add/index.ts [hunk 7]
```diff
@@ -76,21 +82,28 @@ export function ngAddPackages(packages: string[], options?: NgAddPackageOptions,
...
           hideOutput: false,
           quiet: false
         } as any));
         await lastValueFrom(context.engine.executePostTasks());
 
-        const ngAddCollection = getNgAddSchema(packageName, context);
-        if (ngAddCollection) {
-          const ngAddOptions = getOptions(ngAddCollection);
-          return () => externalSchematic(packageName, 'ng-add', ngAddOptions)(tree, context);
+        if (!options?.skipNgAddSchemticRun) {
```
Comment (matthieu-crouzet) [inline packages/@o3r/schematics/src/rule-factories/ng-add/index.ts:95]:
```suggestion
        if (!options?.skipNgAddSchematicRun) {
```

File: packages/@o3r/schematics/src/tasks/ng-add/index.ts [hunk 8]
```diff
@@ -26,6 +26,9 @@ export interface NgAddPackageOptions {
    * Type of dependency to install
    */
   dependencyType?: NodeDependencyType;
+
+  /** Flag to skip the execution of ng add and only install the package. Used mostly for external packages */
+  skipNgAddSchemticRun?: boolean;
```
Comment (matthieu-crouzet) [inline packages/@o3r/schematics/src/tasks/ng-add/index.ts:31]:
```suggestion
  skipNgAddSchematicRun?: boolean;
```

File: packages/@o3r/styling/schematics/ng-add/index.ts [hunk 9]
```diff
@@ -57,7 +60,7 @@ export function ngAdd(options: NgAddSchematicsSchema): Rule {
             useOtterTheming: undefined
           }
         }),
-        ngAddPeerDependencyPackages(['chokidar'], packageJsonPath, NodeDependencyType.Dev, options, depsInfo.packageName),
+        ngAddPeerDependencyPackages(['chokidar'], packageJsonPath, NodeDependencyType.Dev, {...options, workingDirectory, skipNgAddSchemticRun: true}, depsInfo.packageName),
```
Comment (matthieu-crouzet) [inline packages/@o3r/styling/schematics/ng-add/index.ts:63]:
```suggestion
        ngAddPeerDependencyPackages(['chokidar'], packageJsonPath, NodeDependencyType.Dev, {...options, workingDirectory, skipNgAddSchematicRun: true}, depsInfo.packageName),
```

File: packages/@o3r/testing/schematics/ng-add/index.ts [hunk 10]
```diff
@@ -32,9 +34,10 @@ export function ngAdd(options: NgAddSchematicsSchema): Rule {
           skipConfirmation: true,
           version: depsInfo.packageVersion,
           parentPackageInfo: depsInfo.packageName,
-          dependencyType: dependencyType
+          dependencyType: dependencyType,
+          workingDirectory
         }),
-        ngAddPeerDependencyPackages(['pixelmatch', 'pngjs'], testPackageJsonPath, dependencyType, options),
+        ngAddPeerDependencyPackages(['pixelmatch', 'pngjs'], testPackageJsonPath, dependencyType, {...options, workingDirectory, skipNgAddSchemticRun: true}),
```
Comment (matthieu-crouzet) [inline packages/@o3r/testing/schematics/ng-add/index.ts:40]:
```suggestion
        ngAddPeerDependencyPackages(['pixelmatch', 'pngjs'], testPackageJsonPath, dependencyType, {...options, workingDirectory, skipNgAddSchematicRun: true}),
```

**Review by matthieu-crouzet — DISMISSED** [review]

**Review by matthieu-crouzet — DISMISSED** [review]

**Review by vscaiceanu-1a — DISMISSED** [review]

**Review by matthieu-crouzet — DISMISSED** [review]

File: packages/@o3r/components/schematics/ng-add/index.ts (hunk 1 above)
Comment (mrednic-1A) [reply packages/@o3r/components/schematics/ng-add/index.ts]: Nice typo :) Nice catch :)

File: packages/@o3r/core/schematics/ng-add/project-setup/index.ts [hunk 11]
```diff
@@ -72,16 +73,20 @@ export const prepareProject = (options: NgAddSchematicsSchema) => async (tree: T
     options.generateAzurePipeline ? createAzurePipeline(options, coreSchematicsFolder) : noop,
     addVsCodeRecommendations(['AmadeusITGroup.otter-devtools', 'EditorConfig.EditorConfig', 'angular.ng-template'])
   ];
+
+  const projectDirectory = workspaceProject ? workspaceProject.root : undefined;
```
Comment (kpanot) [inline packages/@o3r/core/schematics/ng-add/project-setup/index.ts:77]:
```suggestion
  const projectDirectory = workspaceProject?.root;
```

File: packages/@o3r/core/schematics/rule-factories/otter-environment/index.ts [hunk 12]
```diff
@@ -116,21 +118,22 @@ export function updateOtterEnvironmentAdapter(
...
   const generateEnvironmentFiles = (tree: Tree, context: SchematicContext) => {
+
     const workspaceProject = getProjectFromTree(tree, options.projectName, 'application');
     if (!workspaceProject) {
       context.logger.error('No application detected, the environment can not be generated');
       return tree;
     }
 
-    if (tree.exists('/src/environments/environment.ts')) {
+    if (tree.exists(nodePath.join(workspaceProject.root, '/src/environments/environment.ts'))) {
```
Comment (kpanot) [inline packages/@o3r/core/schematics/rule-factories/otter-environment/index.ts:128]: why the initial `/` which is a bit misleading?

File: packages/@o3r/schematics/src/rule-factories/ng-add/index.ts [hunk 13]
```diff
@@ -16,6 +17,10 @@ import { getExternalDependenciesVersionRange, getNodeDependencyList, getPackageM
  * @param packageJsonPath path of the package json of the project where they will be installed
  */
 export function ngAddPackages(packages: string[], options?: NgAddPackageOptions, packageJsonPath = '/package.json'): Rule {
+  if (options?.workingDirectory && !packageJsonPath.startsWith(options.workingDirectory)) {
```
Comment (kpanot) [inline packages/@o3r/schematics/src/rule-factories/ng-add/index.ts:20]: I suppose that this is to convert relative path to absolute on, the `startWith` may not be the best on.
```typescript
if (options?.workingDirectory && !path.isAbsolute(options.workingDirectory)) {
```
?

File: packages/@o3r/schematics/src/rule-factories/ng-add/index.ts [hunk 14]
```diff
@@ -84,11 +90,16 @@ export function ngAddPackages(packages: string[], options?: NgAddPackageOptions,
     });
     tree.overwrite(packageJsonPath, JSON.stringify(packageJson, null, 2));
 
+    if (options?.skipNgAddSchematicRun) {
+      context.logger.info(`Packages '${packagesToInstall.join(', ')}' were installed. The run of 'ng-add' schematics for these packages is intentionally skipped. You can run them standalone, later.`);
+      return () => tree;
```
Comment (kpanot) [inline packages/@o3r/schematics/src/rule-factories/ng-add/index.ts:95]: why not just `return tree;` ?

File: packages/@o3r/schematics/src/tasks/ng-add/index.ts [hunk 15]
```diff
@@ -26,6 +26,9 @@ export interface NgAddPackageOptions {
    * Type of dependency to install
    */
   dependencyType?: NodeDependencyType;
+
+  /** Flag to skip the execution of ng add and only install the package. Used mostly for external packages */
+  skipNgAddSchematicRun?: boolean;
```
Comment (kpanot) [inline packages/@o3r/schematics/src/tasks/ng-add/index.ts:31]: Is it already part of the `schema.json` (which is not part of this PR)?

File: packages/@o3r/schematics/src/utility/loaders.ts [hunk 16]
```diff
@@ -92,6 +92,18 @@ export function getProjectFromTree(tree: Tree, projectName?: string | null, proj
   return workspaceProject;
 }
 
+/**
+ * Returns the root directory of a project with the given name ( a relative path from the project root)
+ *
+ * @param tree Files tree
+ * @param projectName Name of the project inside the workspace
+ */
+export function getProjectRootDir(tree: Tree, projectName?: string | null) {
```
Comment (kpanot) [inline packages/@o3r/schematics/src/utility/loaders.ts:101]: Maybe it would be cleaner to actually fix `getProjectFromTree` to not throw in case of `!tree.exists('angular.json')` instead of adding this helper (that can to be reproduced to all the fields of the workspace object)

File: packages/@o3r/core/schematics/rule-factories/otter-environment/index.ts (hunk 12 above)
Comment (mrednic-1A) [reply packages/@o3r/core/schematics/rule-factories/otter-environment/index.ts:128]: It''s from copy paste. I'll remove it

File: packages/@o3r/schematics/src/rule-factories/ng-add/index.ts (hunk 13 above)
Comment (mrednic-1A) [reply packages/@o3r/schematics/src/rule-factories/ng-add/index.ts:20]: Not really. It's to add the working directory at the beginning of the **packageJsonPath** if it's not already the there.

File: packages/@o3r/schematics/src/tasks/ng-add/index.ts (hunk 15 above)
Comment (mrednic-1A) [reply packages/@o3r/schematics/src/tasks/ng-add/index.ts:31]: For the moment is just a parameter of the **ngAddPackages** function. 
Passing it as an option will be done in [691](https://github.com/AmadeusITGroup/otter/issues/691)

File: packages/@o3r/schematics/src/rule-factories/ng-add/index.ts (hunk 14 above)
Comment (mrednic-1A) [reply packages/@o3r/schematics/src/rule-factories/ng-add/index.ts:95]: the return type of the parent function forces a Rule

File: packages/@o3r/application/schematics/ng-add/index.ts [hunk 17]
```diff
@@ -1,20 +1,24 @@
 import type { Rule, SchematicContext, Tree } from '@angular-devkit/schematics';
 import { chain } from '@angular-devkit/schematics';
 import * as path from 'node:path';
+import { NgAddSchematicsSchema } from './schema';
```
Comment (vscaiceanu-1a) [inline packages/@o3r/application/schematics/ng-add/index.ts:4]:
```suggestion
import type { NgAddSchematicsSchema } from './schema';
```

File: packages/@o3r/configuration/schematics/ng-add/index.ts [hunk 18]
```diff
@@ -1,11 +1,13 @@
 import { chain, SchematicContext, Tree } from '@angular-devkit/schematics';
 import * as fs from 'node:fs';
 import * as path from 'node:path';
+import { NgAddSchematicsSchema } from './schema';
```
Comment (vscaiceanu-1a) [inline packages/@o3r/configuration/schematics/ng-add/index.ts:4]:
```suggestion
import type { NgAddSchematicsSchema } from './schema';
```

File: packages/@o3r/core/schematics/rule-factories/additional-modules/index.ts [hunk 19]
```diff
@@ -22,7 +22,7 @@ const ngrxStoreDevtoolsDep = '@ngrx/store-devtools';
  * @param options.projectName
  * @param _rootPath @see RuleFactory.rootPath
  */
-export function updateAdditionalModules(options: { projectName?: string | null | undefined }, _rootPath: string): Rule {
+export function updateAdditionalModules(options: { projectName?: string | null | undefined; workingDirectory?: string | undefined }, _rootPath: string): Rule {
```
Comment (vscaiceanu-1a) [inline packages/@o3r/core/schematics/rule-factories/additional-modules/index.ts:25]: would be nice to have some TSdoc for this new param

File: packages/@o3r/mobile/schematics/ng-add/index.ts [hunk 20]
```diff
@@ -1,20 +1,22 @@
 import { chain, Rule, SchematicContext, Tree } from '@angular-devkit/schematics';
 import { getProjectDepType } from '@o3r/schematics';
 import * as path from 'node:path';
+import { NgAddSchematicsSchema } from './schema';
 
 /**
  * Add Otter mobile to an Angular Project
  */
-export function ngAdd(): Rule {
+export function ngAdd(options: NgAddSchematicsSchema): Rule {
```
Comment (vscaiceanu-1a) [inline packages/@o3r/mobile/schematics/ng-add/index.ts:9]: missing tsdoc

File: packages/@o3r/schematics/src/utility/modules.ts [hunk 21]
```diff
@@ -20,11 +20,11 @@ import {getProjectFromTree} from './loaders';
  * @param tree File tree
  * @param context Context of the rule
  */
-export function getAppModuleFilePath(tree: Tree, context: SchematicContext) {
-  const workspaceProject = getProjectFromTree(tree, null, 'application');
+export function getAppModuleFilePath(tree: Tree, context: SchematicContext, projectName?: string | null) {
```
Comment (vscaiceanu-1a) [inline packages/@o3r/schematics/src/utility/modules.ts:23]: js doc for projectName

File: packages/@o3r/core/schematics/ng-add/project-setup/index.ts [hunk 22]
```diff
@@ -51,7 +51,8 @@ export const prepareProject = (options: NgAddSchematicsSchema) => async (tree: T
     ...(options.enableLocalization ? ['@o3r/localization'] : []),
     ...(options.enableCustomization ? ['@o3r/components', '@o3r/configuration'] : []),
     ...(options.enableStorybook ? ['@o3r/storybook'] : []),
-    ...(installOtterLinter ? ['@o3r/eslint-config-otter'] : [])
+    ...(installOtterLinter ? ['@o3r/eslint-config-otter'] : []),
+    ...(projectType ? ['@ama-sdk/core'] : ['@ama-sdk/schematics'])
```
Comment (fpaul-1A) [inline packages/@o3r/core/schematics/ng-add/project-setup/index.ts:55]: should this really comes with `@o3r/core` ?
I think this should not be installed by default, maybe it can be included in a preset ?

File: packages/@o3r/core/schematics/rule-factories/additional-modules/index.ts [hunk 23]
```diff
@@ -20,9 +20,10 @@ const ngrxStoreDevtoolsDep = '@ngrx/store-devtools';
  *
  * @param options @see RuleFactory.options
  * @param options.projectName
+ * @param options.workingDirectory the directory where tu execute the rule factory
```
Comment (fpaul-1A) [inline packages/@o3r/core/schematics/rule-factories/additional-modules/index.ts:23]:
```suggestion
 * @param options.workingDirectory the directory where to execute the rule factory
```

File: packages/@o3r/core/schematics/ng-add/project-setup/index.ts (hunk 22 above)
Comment (mrednic-1A) [reply packages/@o3r/core/schematics/ng-add/project-setup/index.ts]: updated. as discussed I called the helper function to retrieve the peer deps instead of hardcoding them.

File: packages/@o3r/core/schematics/rule-factories/otter-environment/index.ts [hunk 24]
```diff
@@ -116,21 +118,22 @@ export function updateOtterEnvironmentAdapter(
...
   const generateEnvironmentFiles = (tree: Tree, context: SchematicContext) => {
+
     const workspaceProject = getProjectFromTree(tree, options.projectName, 'application');
     if (!workspaceProject) {
       context.logger.error('No application detected, the environment can not be generated');
       return tree;
     }
 
-    if (tree.exists('/src/environments/environment.ts')) {
+    if (tree.exists(nodePath.join(workspaceProject.root, 'src/environments/environment.ts'))) {
```
[the remark this answers is not in the document]
Comment (mrednic-1A) [reply packages/@o3r/core/schematics/rule-factories/otter-environment/index.ts]: updated :)

File: packages/@o3r/schematics/src/utility/modules.ts (hunk 21 above)
Comment (mrednic-1A) [reply packages/@o3r/schematics/src/utility/modules.ts:24]: updated

File: packages/@o3r/application/schematics/ng-add/index.ts (hunk 17 above)
Comment (mrednic-1A) [reply packages/@o3r/application/schematics/ng-add/index.ts]: done

**Review by fpaul-1A — APPROVED** [review]

File: packages/@o3r/core/schematics/rule-factories/otter-environment/index.ts [hunk 25]
```diff
@@ -3,6 +3,7 @@ import type { PackageManager } from '@angular/cli/lib/config/workspace-schema';
 import generateEnvironments from '@schematics/angular/environments/index';
 import * as ts from 'typescript';
 import { getPackageManager, getProjectFromTree, OTTER_ITEM_TYPES, readAngularJson, readPackageJson, registerCollectionSchematics, TYPES_DEFAULT_FOLDER } from '@o3r/schematics';
+import * as nodePath from 'node:path';
```
Comment (kpanot) [inline packages/@o3r/core/schematics/rule-factories/otter-environment/index.ts:6]: Could be directly

```suggestion
import { posix } from 'node:path';
```

File: packages/@o3r/schematics/src/utility/loaders.ts [hunk 26]
```diff
@@ -73,7 +73,14 @@ export function readPackageJson(tree: Tree, workspaceProject: WorkspaceProject)
...
  */
 export function getProjectFromTree(tree: Tree, projectName?: string | null, projectType?: 'application' | 'library'): WorkspaceProject & { name: string } | undefined {
-  const workspace = readAngularJson(tree);
+
+  let workspace: WorkspaceSchema;
+  try {
+    workspace = readAngularJson(tree);
+  } catch {
+    return undefined;
+  }
```
Comment (kpanot) [inline packages/@o3r/schematics/src/utility/loaders.ts:82]: use `getWorkspaceConfig` instead

```suggestion
  const workspace = getWorkspaceConfig(tree);
  if (!workspace) {
    return undefined;
  }
```

Comment (mrednic-1A) [reply packages/@o3r/schematics/src/utility/loaders.ts:82]: indeed :)

**Review by vscaiceanu-1a — APPROVED** [review]

**Review by fpaul-1A — APPROVED** [review]

**Review by kpanot — APPROVED** [review]
